### ISSUE TITLE
[Crafting] Spawn/Sync Blueprint Fragments

### ISSUE THREAD

**Issue by MadMaxOfYore** [SubnauticaNitrox/Nitrox#161]
It appears that blueprint fragments are not spawning in the game. Need to determine if this is a bug or if it is just something we have not gotten to yet before we can set priority.

**Comment by Alberto247** [comment]
![immagine](https://user-images.githubusercontent.com/9718334/36227917-e57aded4-11d2-11e8-8ffa-5fee5eea08a0.png)
I've logged all entities spawned by the server, and blueprints/fragments are spawned. They look like a little dark cube, but they can be scanned and they work fine.
Next step is finding why they look like this.

**Comment by MadMaxOfYore** [comment]
Sounds like a good candidate for a PR if you ask me ;)

**Comment by Alberto247** [comment]
![immagine](https://user-images.githubusercontent.com/9718334/36386818-60b19940-1597-11e8-9329-1e4643af8a98.png)
The issue looks like a naming issue as the TechType is SeamothFragment(this is just an example) and inside prefabs.db it's called Seamoth_Fragment.
However, I tried in a fresh installation of the game, without any patch, to give myself the item seamothfragment through the console and it dropped me the same black cube, which is the default aspect of an item not found in CraftData.InstantiateFromPrefab.
Not sure if this is a game issue or somehow the server is using the old TechType, which is now called SeamothFragment_old.
What's more if you use the command "spawn seamothfragment" it tells you that it could not find the prefab for TechType seamothfragment, however if you use the command "spawn seamoth_fragment" it tells you that the TechType seamoth_fragment has not been found.
Things are working just fine with some fragments such as the seaglidefragment, but not for all of them(see cyclops hull fragments).
The logs, when you spawn the seamothfragments says "Error: Could not find prefab class id for tech type SeamothFragment. Probably missing from CraftData.cs".
Not so sure if this is a Nitrox's issue or a Subnautica's issue. Any ideas?
I'm now investigating the fragments whose name is fine and which still don't spawn at all.

**Comment by MarijnS95** [comment]
### Assumption:

Fragments have different models (prefabs) scattered around the world. Is it possible that we need to (randomly) choose one of these prefab models and spawn it. It looks like `ResourceTracker` may be related.

**Comment by Alberto247** [comment]
I'll check that later. Thank you for the help.

**Comment by Alberto247** [comment]
I found that by creating the gameObject through the same procedure used in EntitySlotPlaceholder.Spawn() the fragments are created with the correct random prefab. However, after adding the correct Guid and checking that it has been registered in the UniqueIdentifier class through UniqueIdentifier.Register(), it gets unregistered by UniqueIdentifier.Unregister() with apparently no reason. However the model remains in the world, and can be still interacted with, the only problem is that animated entities don't move as UpdateEntityPosition() can't find the gameObject from the Guid anymore.
My current solution was to create two procedures to create the gameObject, one for the fragments and one for all the other entities, but of course this is not the best solution.
I'm sure i'm missing something important, but I can't get to the answer.
Sorry for bothering so much, i'm still learning C#, hope i'm not causing too much trouble.


```
GameObject virtualEntityPrefab = VirtualEntitiesManager.GetVirtualEntityPrefab();
virtualEntityPrefab.SetActive(false);
if (PDAScanner.IsFragment(entity.TechType))
{
         GameObject gameObject = UnityEngine.Object.Instantiate<GameObject>(virtualEntityPrefab, vector, quaternion);
         gameObject.transform.SetParent(null, false);
         gameObject.transform.localScale = entity.LocalScale;
         gameObject.GetComponent<VirtualPrefabIdentifier>().ClassId = entity.ClassId;
         LargeWorldEntity component = gameObject.GetComponent<LargeWorldEntity>();
         component.cellLevel = entity.CellLevel;
         if (LargeWorld.main != null)
         {
                 LargeWorld.main.streamer.cellManager.RegisterEntity(component);
                 LargeWorldEntity.Register(gameObject);
         }
         GuidHelper.SetNewGuid(gameObject, entity.Guid);
         gameObject.active = true;
}
else
{
         GameObject gameObject = CraftData.InstantiateFromPrefab(entity.TechType);
         gameObject.transform.position = vector;
         gameObject.transform.localRotation = quaternion;
         GuidHelper.SetNewGuid(gameObject, entity.Guid);
         gameObject.active = true;
         LargeWorldEntity.Register(gameObject);
         CrafterLogic.NotifyCraftEnd(gameObject, entity.TechType);
}
```

**Comment by Sunrunner37** [comment]
@Alberto247 You have been extremely helpful - thanks a ton for reporting this bugs.  I introduced some new custom spawner code that will be able to help us with this.  I'll tackle this issue soon.

**Comment by Sunrunner37** [comment]
@Alberto247 - This has been resolved in the latest commit (793ba01).  Some entities have to leverage the firstpass prefab database instead of the techtype prefab database.  The code that you wrote forces the fragments to use the VirtualEntity which, in the background, uses the firstpass prefab database.  However, VirtualEntity spawns a new game object in the first frame - this is why your GUID was not matching up because you were not getting it assigned to the actual fragment object.  To solve this problem, I decided to use the firstpass prefab database directly
